### PR TITLE
Ac/beamforming delay and sum

### PR DESCRIPTION
**Pull request by eljuguetero**

### CODE REVIEW

File: cpp/driver/circular_queue.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  /*
+    Resize buffer for circular fifo
+  */
+  bool Resize(int N) {
+    if (N < 0) {
+      return false;
+    }
+
+    if (fifo_) {
```
Comment (arhuaco) [inline cpp/driver/circular_queue.h]: It's OK to delete null pointers. It's noop.

File: cpp/driver/circular_queue.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  ~CircularQueue() {
+    if (fifo_) {
+      delete[] fifo_;
+    }
+  }
+
+  /*
+    Resize buffer for circular fifo
+  */
+  bool Resize(int N) {
```
Comment (arhuaco) [inline cpp/driver/circular_queue.h]: new_size instead of N?

File: cpp/driver/circular_queue.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    if (fifo_) {
+      delete[] fifo_;
+    }
+
+    fifo_size_ = N + 1;
+    pointer_ = 0;
+    fifo_ = new T[fifo_size_];
+    return (fifo_ != NULL);
```
Comment (arhuaco) [inline cpp/driver/circular_queue.h]: ( .. ) not needed, you can just use the expression inside of the parens.

File: cpp/driver/circular_queue.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+#ifndef CPP_DRIVER_CIRCULAR_QUEUE_
+#define CPP_DRIVER_CIRCULAR_QUEUE_
+
+namespace matrix_hal {
+
+template <class T>
+class CircularQueue {
+ public:
+  CircularQueue() : fifo_size_(0), pointer_(0), fifo_(NULL) {}
```
Comment (arhuaco) [inline cpp/driver/circular_queue.h]: nullptr instead of NULL.

File: cpp/driver/circular_queue.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    return (fifo_ != NULL);
+  }
+
+  /*
+    Simultaneous Push/Pop method
+  */
+  T PushPop(const T& data) {
+    if (fifo_size_ > 1) {
+      fifo_[(pointer_ + fifo_size_ - 1) % fifo_size_] = data;
+      T& ret = fifo_[pointer_];
```
Comment (arhuaco) [inline cpp/driver/circular_queue.h]: Just a note, not asking for a change: I think you can gain a position if you take the reference before overwriting (with some changes).

File: cpp/driver/circular_queue.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    fifo_size_ = N + 1;
+    pointer_ = 0;
+    fifo_ = new T[fifo_size_];
+    return (fifo_ != NULL);
+  }
+
+  /*
+    Simultaneous Push/Pop method
+  */
+  T PushPop(const T& data) {
```
Comment (arhuaco) [inline cpp/driver/circular_queue.h]: Why a reference and not a pointer? The issue of taking references is that they might not exist when you try to dereference them. Same as pointers, but with pointers it's explicit. I find this weird with references.

File: cpp/driver/microphone_array.cpp [hunk 7]
```diff
@@ -52,9 +68,62 @@ bool MicrophoneArray::Read() {
       return false;
     }
 
-    for (auto& data : raw_data_) data = data * gain_;
+    for (uint32_t s = 0; s < NumberOfSamples(); s++) {
+      int sum = 0;
+      for (uint16_t c = 0; c < kMicrophoneChannels; c++) {
+        delayed_data_[s * kMicrophoneChannels + c] =
+            fifos_[c].PushPop(raw_data_[s * kMicrophoneChannels + c]) * gain_;
```
Comment (arhuaco) [inline cpp/driver/microphone_array.cpp]: Here. It's hard to guess that PushPop is getting a reference of raw_data[ ... ].

A pointer makes it easier.

File: cpp/driver/microphone_array.cpp [hunk 8]
```diff
@@ -52,9 +68,62 @@ bool MicrophoneArray::Read() {
...
+
+        sum = sum + delayed_data_[s * kMicrophoneChannels + c];
```
Comment (arhuaco) [inline cpp/driver/microphone_array.cpp]: sum +=

File: cpp/driver/microphone_array.cpp [hunk 9]
```diff
@@ -52,9 +66,55 @@ bool MicrophoneArray::Read() {
...
+      }
+
+      if (sum < INT16_MIN) {
+        beamformed_[s] = INT16_MIN;
+      } else if (sum > INT16_MAX) {
```
Comment (arhuaco) [inline cpp/driver/microphone_array.cpp]: #include <algorithm>

    sum = std::max(sum, INT16_MIN);
    sum = std::min(sum, INT16_MAX);
    beamformed_[s] = sum;

or:

    beamformed_[s] = std::min(INT16_MAX, std::min(sum, INT16_MAX));

File: cpp/driver/microphone_array.cpp [hunk 10]
```diff
@@ -52,9 +66,55 @@ bool MicrophoneArray::Read() {
...
+  // sorted distances from source position to each microphone
+  for (int c = 0; c < kMicrophoneChannels; c++) {
+    float distance = std::sqrt(std::pow(micarray_location[c][0] - x, 2.0) +
+                               std::pow(micarray_location[c][1] - y, 2.0) +
+                               std::pow(z, 2.0));
+    distance_map[distance] = c;
+  }
+
+  // fifo resize for delay compensation
+  float min_distance = distance_map.begin()->first;
```
Comment (arhuaco) [inline cpp/driver/microphone_array.cpp]: const

File: cpp/driver/microphone_array.cpp (hunk 7 above)
Comment (arhuaco) [inline cpp/driver/microphone_array.cpp]: Not asking for a change...

File: cpp/driver/microphone_array.cpp (hunk 9 above)
Comment (eljuguetero) [inline cpp/driver/microphone_array.cpp]: beamformed_[s]  = std::min(INT16_MAX, std::max(sum, INT16_MIN));

File: cpp/driver/microphone_array.cpp (hunk 8 above)
Comment (eljuguetero) [inline cpp/driver/microphone_array.cpp]: Reluctantly I will

**Pull request merged**